File: Shared/Resources/quicksort.bundle/quicksort.rb
def partition(array, first, last)
  pivot = array[last]
  pIndex = first
  i = first
  while i < last
    if array[i].to_i <= pivot.to_i
      array[i], array[pIndex] = array[pIndex], array[i]
      pIndex += 1
    end
    i += 1
  end
  array[pIndex], array[last] = array[last], array[pIndex]
  return pIndex
end

def quick_sort(array, first, last)
  if first < last
    j = partition(array, first, last)
    quick_sort(array, first, j-1)
    quick_sort(array, j+1, last)
  end
  return array
end

def sort(array)
  return quick_sort(array, 0, array.length() - 1)
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array

File: Shared/Resources/quicksort.bundle/quicksort.swift
import Foundation

func partition(
  _ array: inout [Int],
  _ begin: Int,
  _ end: Int
) -> Int {
  let pivot = array[end]
  var i = begin - 1
  for j in begin..<end {
    if array[j] <= pivot {
      i += 1
      array.swapAt(i, j)
    }
  }
  array.swapAt(i + 1, end)
  return i + 1
}

func quickSort(_ array: inout [Int], _ begin: Int, _ end: Int) {
  if begin < end {
    let partitionIndex = partition(&array, begin, end)
    quickSort(&array, begin, partitionIndex - 1)
    quickSort(&array, partitionIndex + 1, end)
  }
}

func sort(_ array: inout [Int]) {
  quickSort(&array, 0, array.count - 1)
}

var array: [Int] = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(&array)
print(array)